_posts/2021-03-23-keccak-optimizer-bug.md

```diff
@@ -0,0 +1,89 @@
+date:   2021-03-23
+author: Solidity Team
+category: Security Alerts
+---
+
+On March 20, 2021, a bug in Solidity's bytecode optimizer was found. The bug is fixed with [version
+0.8.3](https://github.com/ethereum/solidity/releases/tag/v0.8.3) released on March 23, 2021. **The
+bug is present in all prior versions of Solidity.**
+
+We assigned the bug a severity level of "medium".
+
+## Technical Details of the Bug
+
+**Summary**: The bytecode optimizer incorrectly re-used previously evaluated Keccak-256 hashes. You
+are unlikely to be affected if you do not compute Keccak-256 hashes in inline assembly.
+
+Solidity's bytecode optimizer has a step that can compute Keccak-256 hashes, if the contents of the
+memory are known during compilation time. This step also has a mechanism to determine that two
+Keccak-256 hashes are equal even if the values in memory are not known during compile time. This
+mechanism had a bug where, Keccak-256 of the same memory content, but different sizes were
+considered equal. The following is a example that illustrates the bug:
+
+```solidity
+contract C {
+  uint[] data;
+
+  function bug() public{
+    assembly {
+      // Sets data.length = 2
+      sstore(0, 2)
+      mstore(0, 0)
+      // data[0] = 234
+      sstore(keccak256(0, 32), 234)
+      // A bug in the caching mechanism previously caused keccak256(0, 23) to be the same as
+      // keccak256(0, 32), leading to data[1] == 123` being true.
+      sstore(add(keccak256(0, 23), 1), 123)
+    }
+  }
+}
+```
+
+In the above example, both the Keccak-256 hashes can be computed at compile time. Here are two other
+cases where the hashes cannot be computed at compile time, yet assuming equality breaks the
+semantics:
+
+```solidity
+contract C {
+  function bug(string memory s) public returns (bool ret) {
+    assembly {
+      let a := keccak256(s, 32)
+      let b := keccak256(s, 8)
+      // Here `a` and `b` were considered equal,
+      // leading to `ret` being incorrectly set to true.
+      ret := eq(a, b)
+    }
+  }
+}
+```
+
+```solidity
+contract C {
+  function bug() public view returns (bool ret) {
+    assembly {
+      let x := calldataload(0)
+      mstore(0, x)
+      mstore(0x20, x)
+      let a := keccak256(0, 4)
+      // even though the memory location is different,
+      // the 32-byte content is the same.
+      let b := keccak256(0x20, 8)
+      // Here `a` and `b` were considered equal,
+      // leading to `ret` being incorrectly set to true.
+      ret := eq(a, b)
+    }
+  }
+}
+```
+
+Specifically, ``keccak256(mpos1, length1)`` and ``keccak256(mpos2, length2)`` in some cases were
+considered equal if ``length1`` and ``length2``, when rounded up to nearest multiple of 32 were the
+same, and when the memory contents at ``mpos1`` and ``mpos2`` can be deduced to be equal.
+
+You maybe affected if you compute multiple Keccak-256 hashes of the same content, but with different
+length inside inline assembly, and enabled the optimizer.
+
```

Comment:
We should also add a section that explains how we arrived at the severity classification.